### PR TITLE
sam:d: port: complete port (define, configuration and access)

### PR DESCRIPTION
**Pull request by trabucayre**
Tested with Feather M0 based on SAMD21 with success

### CODE REVIEW

**Comment by grevaillot** [comment]
nice !

I had a project cancelled at work on samr34, so i started some support for d/l/r chips last summer. I'll have a review, but i think you can pick some bits - not much time currently to continue work in it :)

https://github.com/libopencm3/libopencm3/compare/master...grevaillot:wip/samlr

File: include/libopencm3/sam/d/port.h [hunk 1]
```diff
@@ -66,3 +134,216 @@
...
+	GPIO_MODE_INPUT_PULLUP,
+	GPIO_MODE_AF
+};
+/**@}*/
+
+/* --- Function prototypes ------------------------------------------------- */
+
+BEGIN_DECLS
+
+void gpio_mode_setup(uint32_t gpioport, enum gpio_mode mode, uint32_t gpios);
```
Comment (grevaillot) [inline include/libopencm3/sam/d/port.h:261]: most most chips here (except efm32 it seems) are are using mode setup with pull/down parameter :

> void gpio_mode_setup(uint32_t gpioport, uint8_t mode, uint8_t pull_up_down, uint16_t gpios)

maybe we could stick to that, and keep the same gpio_mode enum ?

File: include/libopencm3/sam/d/port.h [hunk 2]
```diff
@@ -66,3 +134,216 @@
...
+
+/** @brief Set a group of Pins (atomic)
+ *
+ * Set a group of Pins for the given port. this is an atomic operation.
+ *
+ * @param[in] gpioport port register address base @ref port_reg_base
+ * @param[in] gpios @ref gpio_defines. Any combinaison of pins may be
+ *          specified by OR'ing then together.
+ */
+static inline void gpio_set(uint32_t gpioport, uint32_t gpios)
```
Comment (grevaillot) [inline include/libopencm3/sam/d/port.h:276]: i'm not sure it is really usefull to inline these and keep them into headers -  compiler will do the job if needed, no ?

Comment (karlp) [reply include/libopencm3/sam/d/port.h:276]: I'll go further and just say, "don't add static inline functions to the headers"

Comment (trabucayre) [reply include/libopencm3/sam/d/port.h]: Thanks all for review. I've updated in consequence

**Comment by karlp** [comment]
KARL: note to self, compare with _old_ samd stuff, see if anything overlaps

**Comment by karlp** [comment]
this has merged in d11680638b with a set of followup commits on style and doxygen.  thanks for this!

**Pull request closed without merging**